nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py

```diff
@@ -0,0 +1,327 @@
+from nemo.core.classes.common import typecheck
+from nemo.utils import logging
+
+__all__ = ['MTEncDecModel']
+
+
+class MTEncDecModel(EncDecNLPModel):
+    """
+    Encoder-decoder machine translation model.
+    """
+
+    def __init__(self, cfg: MTEncDecModelConfig, trainer: Trainer = None):
+
+        self.setup_enc_dec_tokenizers(cfg)
+
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.encoder = instantiate(cfg.encoder)
+        self.encoder_embedding = TransformerEmbedding(
+            vocab_size=self.encoder_vocab_size,
+            hidden_size=cfg.encoder.hidden_size,
+            max_sequence_length=cfg.encoder_embedding.max_sequence_length,
+            num_token_types=cfg.encoder_embedding.num_token_types,
+            embedding_dropout=cfg.encoder_embedding.embedding_dropout,
+            learn_positional_encodings=cfg.encoder_embedding.learn_positional_encodings,
+        )
+        self.decoder = instantiate(cfg.decoder)
+        self.decoder_embedding = TransformerEmbedding(
+            vocab_size=self.decoder_vocab_size,
+            hidden_size=cfg.decoder.hidden_size,
+            max_sequence_length=cfg.decoder_embedding.max_sequence_length,
+            num_token_types=cfg.decoder_embedding.num_token_types,
+            embedding_dropout=cfg.decoder_embedding.embedding_dropout,
+            learn_positional_encodings=cfg.decoder_embedding.learn_positional_encodings,
+        )
+
+        self.log_softmax = TokenClassifier(
+            hidden_size=cfg.decoder.hidden_size,
+            num_classes=self.decoder_vocab_size,
+            activation=cfg.head.activation,
+            log_softmax=cfg.head.log_softmax,
+            dropout=cfg.head.dropout,
+            use_transformer_init=cfg.head.use_transformer_init,
+        )
+
+        self.beam_search = BeamSearchSequenceGenerator(
+            embedding=self.decoder_embedding,
+            decoder=self.decoder,
+            log_softmax=self.log_softmax,
+            max_sequence_length=cfg.decoder_embedding.max_sequence_length,
+            beam_size=cfg.beam_size,
+            bos=self.decoder_tokenizer.bos_id,
+            pad=self.decoder_tokenizer.pad_id,
+            eos=self.decoder_tokenizer.eos_id,
+            len_pen=cfg.len_pen,
+            max_delta_length=cfg.max_generation_delta,
+        )
+
+        # tie weights of embedding and softmax matrices
+        self.log_softmax.mlp.layer0.weight = self.decoder_embedding.token_embedding.weight
+
+        # TODO: encoder and decoder with different hidden size?
+        std_init_range = 1 / cfg.encoder.hidden_size ** 0.5
+        self.apply(lambda module: transformer_weights_init(module, std_init_range))
+
+        self.loss_fn = SmoothedCrossEntropyLoss(
+            pad_id=self.decoder_tokenizer.pad_id, label_smoothing=cfg.label_smoothing
+        )
+
+        # Optimizer setup needs to happen after all model weights are ready
+        self.setup_optimization(cfg.optim)
```

Comment:
This is done by PTL properly just prior to training, not necessary to set explicitly in constructor